### PR TITLE
untrusted peer downloader validation

### PR DESCRIPTION
**Pull request by b00ris**
closes https://github.com/status-im/go-ethereum/issues/63

### CODE REVIEW

File: les/fetcher.go [hunk 1]
```diff
@@ -722,17 +724,25 @@ func (f *lightFetcher) checkSyncedHeaders(p *peer) {
...
+		h, unapprovedHashes = f.lastValidTrieNode(p)
 		f.chain.Rollback(unapprovedHashes)
 	}
 
+	n := fp.lastAnnounced
 	var td *big.Int
+	trustedHeaderExisted := false
 	for n != nil {
-		if td = f.chain.GetTd(n.hash, n.number); td != nil {
+		if n.hash == h.Hash() {
```
Comment (JekaMas) [inline les/fetcher.go]: Could you add a few comments here?

File: les/fetcher.go [hunk 2]
```diff
@@ -523,6 +524,7 @@ func (f *lightFetcher) newFetcherDistReqForSync(bestHash common.Hash) *distReq {
 			go func() {
 				p := dp.(*peer)
 				p.Log().Debug("Synchronisation started")
+				f.lastTrustedHeader = f.chain.CurrentHeader()
```
Comment (themue) [inline les/fetcher.go]: Is it safe setting this shared field asynchronously in goroutines?

File: les/fetcher.go [hunk 3]
```diff
@@ -749,19 +759,20 @@ func (f *lightFetcher) checkSyncedHeaders(p *peer) {
...
-func (f *lightFetcher) lastValidTrieNode(p *peer) (*fetcherTreeNode, []common.Hash) {
+func (f *lightFetcher) lastValidTrieNode(p *peer) (*types.Header, []common.Hash) {
 	unapprovedHashes := make([]common.Hash, 0)
 
-	fp := f.peers[p]
-	last := fp.lastAnnounced
-	for last != nil {
-		if f.isTrustedHash(last.hash) {
+	current := f.chain.CurrentHeader()
+	for current != nil || current != f.lastTrustedHeader {
```
Comment (themue) [inline les/fetcher.go]: Here is where I meant the comparison of the `lastTrustedHeader` set async in a goroutine.

**Comment by b00ris** [comment]
@themue fixed

**Comment by themue** [comment]
@b00ris Seen my comment regarding setting it inside the goroutine?

**Pull request merged**